src/policy/descriptor.rs

```diff
@@ -0,0 +1,173 @@
+        let spend_info = builder
+            .finalize(&secp, internal_key.to_x_only_pubkey())
+            .expect("constant tree");
+
+        Ok(Self {
+            internal_key,
+            spend_info,
+            policy,
+            cmr,
+        })
+    }
+
+    /// Create a new descriptor from the given policy which will become a single tap leaf
+    ///
+    /// The internal key is set to a constant that is provably not spendable
+    pub fn single_leaf(policy: Policy<Pk>) -> Result<Self, crate::Error> {
+        let bytes = Vec::<u8>::from(UNSPENDABLE_PUBLIC_KEY);
+        let internal_key = Pk::from_32_bytes(&bytes);
+        Self::new(internal_key, policy)
+    }
+
+    /// Return the internal key
+    pub fn internal_key(&self) -> &Pk {
+        &self.internal_key
+    }
+
+    /// Return the spend data
+    pub fn spend_info(&self) -> &TaprootSpendInfo {
+        &self.spend_info
+    }
+
+    /// Return the script pubkey
+    pub fn script_pubkey(&self) -> elements::Script {
+        let output_key = self.spend_info().output_key();
+        let builder = elements::script::Builder::new();
+        builder
+            .push_opcode(elements::opcodes::all::OP_PUSHNUM_1)
+            .push_slice(&output_key.as_inner().serialize())
+            .into_script()
+    }
+
+    /// Return the Elements address on the given network
+    pub fn address(&self, params: &'static elements::AddressParams) -> elements::Address {
+        let output_key = self.spend_info().output_key();
+        elements::Address::p2tr_tweaked(output_key, None, params)
+    }
+
+    /// Return the CMR of the program inside the single tap leaf
+    pub fn cmr(&self) -> Cmr {
+        self.cmr
+    }
+
+    /// Return the single tap leaf
+    pub fn leaf(&self) -> (elements::Script, LeafVersion) {
+        let script = elements::Script::from(Vec::from(self.cmr.as_ref()));
+        let version = leaf_version();
+        (script, version)
+    }
+
+    /// Return a satisfying non-malleable witness and script sig with minimum weight
+    /// to spend an output controlled by the given descriptor if it is possible to satisfy
+    /// given the `satisfier`
+    pub fn get_satisfaction(
+        &self,
+        satisfier: &PolicySatisfier<Pk>,
+    ) -> Result<(Vec<Vec<u8>>, elements::Script), policy::Error> {
+        let program = self
+            .policy
+            .satisfy(satisfier)
+            .ok_or(policy::Error::CouldNotSatisfy)?;
+
+        // Uncomment code below for sanity check
+        // that program successfully runs on Bit Machine
+        // let mut mac = crate::exec::BitMachine::for_program(&program);
+        // let _output = mac
+        //     .exec(&program, &satisfier.env)
+        //     .expect("sanity check");
+
+        let program_and_witness_bytes = program.encode_to_vec();
+        let cmr_bytes = Vec::from(program.cmr.as_ref());
+
+        // FIXME: Should env be public?
+        let control_block = satisfier.env.control_block();
```

Comment:
In fe24e0c692af3361be7a2476d34baca666c45903:

Is this `FIXME` asking whether `env` should be a field or an accessor method on `PolicySatisfier`? I think it should be a method.

Reply:
I made `env` private. The other members `keys` and `preimages` are still open, but we might want some sort of integration with `HashMap<Pk, Sig>` etc. which implement the `Satisfier` trait, so I will leave them for later.

Reply:
Why make only one member private? Either this a POD type and everything should be public or it is an encapsulated type and none should be.

In any case, how do are you supposed to construct this object now?